llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -2103,6 +2103,22 @@ static Value *simplifyAndInst(Value *Op0, Value *Op1, const SimplifyQuery &Q,
       isKnownToBeAPowerOfTwo(Op0, Q.DL, /*OrZero*/ true, 0, Q.AC, Q.CxtI, Q.DT))
     return Constant::getNullValue(Op0->getType());
 
+  // (x << N) & ((x << M) - 1) --> 0, where x is known to be a power of 2 and
+  // M <= N.
```

Comment:
This pattern is basically the same as the `and 2^x-1, 2^C --> 0 where x <= C` one above -- but unfortunately I don't see any way to combine them with the analysis helpers we currently have available. So I think your approach of just adding this extra pattern is fine.

Reply:
Yes, I realized it was very similar, so put my code just after. An 'm_Maybe' would be nice, but as you say it doesn't exist yet.